### ISSUE TITLE
[ML] Verify generalization error estimation for multi-step learning 

### ISSUE THREAD

**Issue by valeriy42** [elastic/ml-cpp#2155]
Cross-validation error for hyperparameter estimation should estimate the generalization error (aka test error on the holdout dataset). This works as long as the test and training datasets have identical distributions. For instance, the training set distribution will become skewed when we use disagreement sampling or residual-based sampling for finding update samples.

Implement a test that verifies that the test error does not increase when we do multi-step learning with a random sampling of the update data (distributions of the training data and test data are the same).

## Why is it important?

We optimize hyperparameters, assuming that the validation error matches the test error. 

Suppose the assumption doesn't hold, and we observe the multi-step learning to overfit. In that case, we don't know if the problem is in the algorithm or the representation of the generalization error during training.

**Comment by valeriy42** [comment]
I looked at the past experiments with random sampling and it seems that the convergence of the test error depends on the seed.

![image](https://user-images.githubusercontent.com/1292899/145368860-6b5a0e79-e243-493f-9440-88dbdd0849c0.png)

Those experiment runs were done with hyperparameter scaling
![image](https://user-images.githubusercontent.com/1292899/145369620-355a826f-8b43-4f03-9929-459c095f64c0.png)

**Comment by valeriy42** [comment]
I run a more extensive experiment on the current feature branch. The results confirm that although the test error sometimes goes up on individual runs, on average, it is always not increasing.

![image](https://user-images.githubusercontent.com/1292899/148400838-31cefb9f-e755-40a6-b06d-ced9bfb7b091.png)

![image](https://user-images.githubusercontent.com/1292899/148400896-3d36d04f-5c32-4c64-94fb-39492c106168.png)

See experiments with the comment "generalization error multiple seeds with parameter ranges" for more details.

**Comment by valeriy42** [comment]
Resolved in #2156
